crates/bevy_ecs/hecs/src/query.rs

```diff
@@ -346,208 +358,304 @@ pub struct FetchOr<T>(T);
-        self.0 = NonNull::new_unchecked(value.add(1));
-        Added { value: &*value }
+        let value = self.0.next();
+        Added { value }
     }
 }
 
-/// Query transformer that retrieves components of type `T` that have either been mutated or added since the start of the frame.
-pub struct Changed<'a, T> {
-    value: &'a T,
+/// Query transformer that retrieves components of type `T` that have either been mutated or Changed since the start of the frame.
+pub struct Changed<'a, Q: Query>
+where
+    Q::Fetch: SingleTypeFetch,
+{
+    value: <Q::Fetch as Fetch<'a>>::Item,
 }
 
-impl<'a, T: Component> Deref for Changed<'a, T> {
-    type Target = T;
+impl<Q: Query> SingleTypeFetch for FetchChanged<Q>
+where
+    Q::Fetch: SingleTypeFetch,
+{
+    type Type = <Q::Fetch as SingleTypeFetch>::Type;
+}
+
+unsafe impl<Q: Query> ReadOnlyFetch for FetchChanged<Q> where
+    Q::Fetch: SingleTypeFetch + ReadOnlyFetch
+{
+}
+
+impl<'a, Q: Query> Deref for Changed<'a, Q>
+where
+    Q::Fetch: SingleTypeFetch,
+{
+    type Target = <Q::Fetch as Fetch<'a>>::Item;
 
     #[inline]
-    fn deref(&self) -> &T {
-        self.value
+    fn deref(&self) -> &Self::Target {
+        &self.value
     }
 }
 
-impl<'a, T: Component> Query for Changed<'a, T> {
-    type Fetch = FetchChanged<T>;
+impl<'a, Q: Query> DerefMut for Changed<'a, Q>
+where
+    Q::Fetch: SingleTypeFetch,
+{
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.value
+    }
+}
+
+impl<'a, Q: Query> Query for Changed<'a, Q>
+where
+    Q::Fetch: SingleTypeFetch,
+{
+    type Fetch = FetchChanged<Q>;
 }
 
 #[doc(hidden)]
-pub struct FetchChanged<T>(NonNull<T>, NonNull<bool>, NonNull<bool>);
-unsafe impl<T> ReadOnlyFetch for FetchChanged<T> {}
+pub struct FetchChanged<Q: Query>(Q::Fetch, NonNull<bool>, NonNull<bool>);
 
-impl<'a, T: Component> Fetch<'a> for FetchChanged<T> {
-    type Item = Changed<'a, T>;
+impl<'a, Q: Query> Fetch<'a> for FetchChanged<Q>
+where
+    Q::Fetch: SingleTypeFetch,
+{
+    type Item = Changed<'a, Q>;
 
     fn access(archetype: &Archetype) -> Option<Access> {
-        if archetype.has::<T>() {
-            Some(Access::Read)
-        } else {
-            None
-        }
+        <Q::Fetch as Fetch<'a>>::access(archetype)
     }
 
     fn borrow(archetype: &Archetype) {
-        archetype.borrow::<T>();
+        <Q::Fetch as Fetch<'a>>::borrow(archetype)
     }
 
     unsafe fn get(archetype: &'a Archetype, offset: usize) -> Option<Self> {
-        archetype
-            .get_with_type_state::<T>()
-            .map(|(components, type_state)| {
-                Self(
-                    NonNull::new_unchecked(components.as_ptr().add(offset)),
-                    NonNull::new_unchecked(type_state.added().as_ptr().add(offset)),
-                    NonNull::new_unchecked(type_state.mutated().as_ptr().add(offset)),
-                )
-            })
+        Some(Self(
+            <Q::Fetch as Fetch<'a>>::get(archetype, offset)?,
```

Comment:
lets eliminate the multiple get_type_state calls here to avoid doing redundant (and relatively expensive) work. Its worth calling out that even after resolving the redundant lookup here, this new approach forces us to do an extra hash for these pointer types (one hash for the Fetch::get() call and one hash for retrieving modified/added state).

We might be able to hack around this by making SingleTypeFetch expose a method that returns the Fetch result and the type state. Its a little cumbersome, but for hot-spot code like this i think its worth doing.

Reply:
That makes sense - it also should avoid the extra derefs if we use SingleTypeFetch::Whatever

I'll probably look into this over the weekend